include/nbl/builtin/hlsl/scan/scheduler.hlsl

```diff
@@ -20,75 +20,88 @@ namespace scan
+            if (((uberWorkgroupIndex/WorkgroupSize) < lastWorkgroupSetIndexForLevel  ? (WorkgroupSize-1u) : schedParams.lastWorkgroupSetCountForLevel[level])==doneCount)
+            {
+                level++;
+            }
+        }
+        level = workgroup::BroadcastFirst(level, accessor);
+        return level == 0 ? false : level == prevLevel; // on level 0 never exit early but on higher levels each workgroup is allowed one operation and exits except if promoted to next level
     }
 
-    uint32_t prevLevel = level;
-    if (workgroup::SubgroupContiguousIndex()==0u)
+    uint32_t levelWorkgroupIndex(NBL_CONST_REF_ARG(uint32_t) level)
     {
-      // The uberWorkgroupIndex is always increasing, even after we switch levels, but for each new level the workgroupSetFinishedIndex must reset
-      const uint32_t workgroupSetFinishedIndex = levelWorkgroupIndex(level) / WorkgroupSize;
-      const uint32_t lastWorkgroupSetIndexForLevel = totalWorkgroupsPerLevel(level) / WorkgroupSize;
-      const uint32_t doneCount = glsl::atomicAdd(scanScratchBuf[0u].data[schedParams.workgroupFinishFlagsOffset[level]+workgroupSetFinishedIndex], 1u);
-      //if ((uberWorkgroupIndex != schedParams.cumulativeWorkgroupCount[level] - 1u ? (WorkgroupSize-1u) : schedParams.lastWorkgroupSetCountForLevel[level])==doneCount)
-      if (((uberWorkgroupIndex/WorkgroupSize) < lastWorkgroupSetIndexForLevel  ? (WorkgroupSize-1u) : schedParams.lastWorkgroupSetCountForLevel[level])==doneCount)
-      {
-        level++;
-      }
+        return uberWorkgroupIndex;
     }
-    level = workgroup::BroadcastFirst(level, accessor);
-    return level == 0 ? false : level == prevLevel; // on level 0 never exit early but on higher levels each workgroup is allowed one operation and exits except if promoted to next level
-  }
 
-  uint32_t levelWorkgroupIndex(const in uint32_t level)
-  {
-      return uberWorkgroupIndex;
-  }
-
-    uint32_t totalWorkgroupsPerLevel(const in uint32_t level)
+    uint32_t totalWorkgroupsPerLevel(NBL_CONST_REF_ARG(uint32_t) level)
+    {
+        return level == 0 
+            ? schedParams.cumulativeWorkgroupCount[level] 
+            : schedParams.cumulativeWorkgroupCount[level] - schedParams.cumulativeWorkgroupCount[level-1u];
+    }
+    
+    void resetWorkgroupsStartedBuffer()
     {
-        return level == 0 ? schedParams.cumulativeWorkgroupCount[level] : schedParams.cumulativeWorkgroupCount[level] - schedParams.cumulativeWorkgroupCount[level-1u];
+        // could do scanScratchBuf[0u].workgroupsStarted[SubgroupContiguousIndex()] = 0u but don't know how many invocations are live during this call
+        if(workgroup::SubgroupContiguousIndex() == 0u)
+        {
+            for(uint32_t i = 0; i < params.topLevel; i++)
+            {
+                scanScratchBuf[0u].workgroupsStarted[i] = 0u;
+            }
+        }
```

Comment:
don't you know the workgroup size?

Reply:
My bad, this is actually not used anymore. I think it was being used inside an `inRage` at some point? Not sure. However we can't reset the `workgroupStarted` buffer within the shader at the end of the Reduce step (which was the initial goal) because it's possible that there are still workgroups that are "alive" where they won't be doing any work but will still increase the `workgroupStarted` buffer and some times we end up doing the reset **before** those WGs exit, ending up with some 1 values instead of all 0s.